File: src/main/java/com/upokecenter/cbor/CBORDataUtilities.java
package com.upokecenter.cbor;
/*
Written in 2013 by Peter O.
Any copyright is dedicated to the Public Domain.
http://creativecommons.org/publicdomain/zero/1.0/
If you like this, you should donate to Peter O.
at: http://upokecenter.com/d/
 */

import com.upokecenter.util.*;

    /**
     * Contains methods useful for reading and writing data, with a focus on CBOR.
     */
  public final class CBORDataUtilities {
private CBORDataUtilities() {
}
    private static final int MaxSafeInt = 214748363;

    /**
     * Parses a number whose format follows the JSON specification. See
     * #ParseJSONNumber(string, integersOnly, parseOnly) for more
     * information.
     * @param str A string to parse.
     * @return A CBOR object that represents the parsed number. Returns null if the
     * parsing fails.
     */
    public static CBORObject ParseJSONNumber(String str) {
      return ParseJSONNumber(str, false, false);
    }

    /**
     * Parses a number whose format follows the JSON specification (RFC 7159).
     * Roughly speaking, a valid number consists of an optional minus sign,
     * one or more digits (starting with 1 to 9 unless the only digit is 0),
     * an optional decimal point with one or more digits, and an optional
     * letter E or e with an optional plus or minus sign and one or more
     * digits (the exponent).
     * @param str A string to parse.
     * @param integersOnly If true, no decimal points or exponents are allowed in
     * the string.
     * @param positiveOnly If true, only positive numbers are allowed (the leading
     * minus is disallowed).
     * @return A CBOR object that represents the parsed number. Returns null if the
     * parsing fails.
     */
    public static CBORObject ParseJSONNumber(
      String str,
      boolean integersOnly,
      boolean positiveOnly) {
      if (((str) == null || (str).length() == 0)) {
        return null;
      }
      int offset = 0;
      boolean negative = false;
      if (str.charAt(0) == '-' && !positiveOnly) {
        negative = true;
        ++offset;
      }
      int mantInt = 0;
      FastInteger mant = null;
      int mantBuffer = 0;
      int mantBufferMult = 1;
      int expBuffer = 0;
      int expBufferMult = 1;
      boolean haveDecimalPoint = false;
      boolean haveDigits = false;
      boolean haveDigitsAfterDecimal = false;
      boolean haveExponent = false;
      int newScaleInt = 0;
      FastInteger newScale = null;
      int i = offset;
      // Ordinary number
      if (i < str.length() && str.charAt(i) == '0') {
        ++i;
        haveDigits = true;
        if (i == str.length()) {
          return CBORObject.FromObject(0);
        }
        if (!integersOnly) {
          if (str.charAt(i) == '.') {
            haveDecimalPoint = true;
            ++i;
          } else if (str.charAt(i) == 'E' || str.charAt(i) == 'e') {
            haveExponent = true;
          } else {
            return null;
          }
        } else {
          return null;
        }
      }
      for (; i < str.length(); ++i) {
        if (str.charAt(i) >= '0' && str.charAt(i) <= '9') {
          int thisdigit = (int)(str.charAt(i) - '0');
          if (mantInt > MaxSafeInt) {
            if (mant == null) {
              mant = new FastInteger(mantInt);
              mantBuffer = thisdigit;
              mantBufferMult = 10;
            } else {
              if (mantBufferMult >= 1000000000) {
                mant.Multiply(mantBufferMult).AddInt(mantBuffer);
                mantBuffer = thisdigit;
                mantBufferMult = 10;
              } else {
                mantBufferMult *= 10;
                mantBuffer = (mantBuffer << 3) + (mantBuffer << 1);
                mantBuffer += thisdigit;
              }
            }
          } else {
            mantInt *= 10;
            mantInt += thisdigit;
          }
          haveDigits = true;
          if (haveDecimalPoint) {
            haveDigitsAfterDecimal = true;
            if (newScaleInt == Integer.MIN_VALUE) {
              newScale = (newScale == null) ? ((new FastInteger(newScaleInt))) : newScale;
              newScale.AddInt(-1);
            } else {
              --newScaleInt;
            }
          }
        } else if (!integersOnly && str.charAt(i) == '.') {
          if (!haveDigits) {
            // no digits before the decimal point
            return null;
          }
          if (haveDecimalPoint) {
            return null;
          }
          haveDecimalPoint = true;
        } else if (!integersOnly && (str.charAt(i) == 'E' || str.charAt(i) == 'e')) {
          haveExponent = true;
          ++i;
          break;
        } else {
          return null;
        }
      }
      if (!haveDigits || (haveDecimalPoint && !haveDigitsAfterDecimal)) {
        return null;
      }
      if (mant != null && (mantBufferMult != 1 || mantBuffer != 0)) {
        mant.Multiply(mantBufferMult).AddInt(mantBuffer);
      }
      if (haveExponent) {
        FastInteger exp = null;
        int expInt = 0;
        offset = 1;
        haveDigits = false;
        if (i == str.length()) {
          return null;
        }
        if (str.charAt(i) == '+' || str.charAt(i) == '-') {
          if (str.charAt(i) == '-') {
            offset = -1;
          }
          ++i;
        }
        for (; i < str.length(); ++i) {
          if (str.charAt(i) >= '0' && str.charAt(i) <= '9') {
            haveDigits = true;
            int thisdigit = (int)(str.charAt(i) - '0');
            if (expInt > MaxSafeInt) {
              if (exp == null) {
                exp = new FastInteger(expInt);
                expBuffer = thisdigit;
                expBufferMult = 10;
              } else {
                if (expBufferMult >= 1000000000) {
                  exp.Multiply(expBufferMult).AddInt(expBuffer);
                  expBuffer = thisdigit;
                  expBufferMult = 10;
                } else {
                  // multiply expBufferMult and expBuffer each by 10
                  expBufferMult = (expBufferMult << 3) + (expBufferMult << 1);
                  expBuffer = (expBuffer << 3) + (expBuffer << 1);
                  expBuffer += thisdigit;
                }
              }
            } else {
              expInt *= 10;
              expInt += thisdigit;
            }
          } else {
            return null;
          }
        }
        if (!haveDigits) {
          return null;
        }
        if (exp != null && (expBufferMult != 1 || expBuffer != 0)) {
          exp.Multiply(expBufferMult).AddInt(expBuffer);
        }
        if (offset >= 0 && newScaleInt == 0 && newScale == null && exp ==
            null) {
          newScaleInt = expInt;
        } else if (exp == null) {
          newScale = (newScale == null) ? ((new FastInteger(newScaleInt))) : newScale;
          if (offset < 0) {
            newScale.SubtractInt(expInt);
          } else if (expInt != 0) {
            newScale.AddInt(expInt);
          }
        } else {
          newScale = (newScale == null) ? ((new FastInteger(newScaleInt))) : newScale;
          if (offset < 0) {
            newScale.Subtract(exp);
          } else {
            newScale.Add(exp);
          }
        }
      }
      if (i != str.length()) {
        // End of the String wasn't reached, so isn't a number
        return null;
      }
      if ((newScale == null && newScaleInt == 0) || (newScale != null &&
                                                     newScale.signum() == 0)) {
        // No fractional part
        if (mant != null && mant.CanFitInInt32()) {
          mantInt = mant.AsInt32();
          mant = null;
        }
        if (mant == null) {
          // NOTE: mantInt can only be positive, so overflow is impossible

          if (negative) {
            mantInt = -mantInt;
          }
          return CBORObject.FromObject(mantInt);
        } else {
          BigInteger bigmant2 = mant.AsBigInteger();
          if (negative) {
            bigmant2 = (bigmant2).negate();
          }
          return CBORObject.FromObject(bigmant2);
        }
      } else {
        BigInteger bigmant = (mant == null) ? (BigInteger.valueOf(mantInt)) :
          mant.AsBigInteger();
        BigInteger bigexp = (newScale == null) ? (BigInteger.valueOf(newScaleInt)) :
          newScale.AsBigInteger();
        if (negative) {
          bigmant = (bigmant).negate();
        }
        return CBORObject.FromObject(ExtendedDecimal.Create(
          bigmant,
          bigexp));
      }
    }
  }

File: CBOR/PeterO/Cbor/CBORDataUtilities.cs
/*
Written by Peter O.
Any copyright to this work is released to the Public Domain.
In case this is not possible, this work is also
licensed under the Unlicense: https://unlicense.org/

 */
using System;
using System.Collections.Generic;
using System.Text;
using PeterO.Numbers;

namespace PeterO.Cbor {
  /// <summary>Contains methods useful for reading and writing data, with
  /// a focus on CBOR.</summary>
  public static class CBORDataUtilities {
    private const string HexAlphabet = "0123456789ABCDEF";

    private const long DoubleNegInfinity = unchecked(0xfffL << 52);
    private const long DoublePosInfinity = unchecked(0x7ffL << 52);

    internal static string ToStringHelper(CBORObject obj, int depth) {
      StringBuilder sb = null;
      CBORType type = obj.Type;
      CBORObject curobject;
      if (obj.IsTagged) {
        if (sb == null) {
          if (type == CBORType.TextString) {
            // The default capacity of StringBuilder may be too small
            // for many strings, so set a suggested capacity
            // explicitly
            string str = obj.AsString();
            sb = new StringBuilder(Math.Min(str.Length, 4096) + 16);
          } else {
            sb = new StringBuilder();
          }
        }
        // Append opening tags if needed
        curobject = obj;
        while (curobject.IsTagged) {
          EInteger ei = curobject.MostOuterTag;
          _ = sb.Append(ei.ToString());
          _ = sb.Append('(');
          curobject = curobject.UntagOne();
        }
      }
      string simvalue;
      switch (type) {
        case CBORType.SimpleValue:
          sb = sb ?? new StringBuilder();
          if (obj.IsUndefined) {
            _ = sb.Append("undefined");
          } else if (obj.IsNull) {
            _ = sb.Append("null");
          } else {
            _ = sb.Append("simple(");
            int thisItemInt = obj.SimpleValue;
            char c;
            if (thisItemInt >= 100) {
              // NOTE: '0'-'9' have ASCII code 0x30-0x39
              c = (char)(0x30 + ((thisItemInt / 100) % 10));
              _ = sb.Append(c);
            }
            if (thisItemInt >= 10) {
              c = (char)(0x30 + ((thisItemInt / 10) % 10));
              _ = sb.Append(c);
              c = (char)(0x30 + (thisItemInt % 10));
            } else {
              c = (char)(0x30 + thisItemInt);
            }
            _ = sb.Append(c);
            _ = sb.Append(')');
          }
          break;
        case CBORType.Boolean:
        case CBORType.Integer:
          simvalue = obj.Untag().ToJSONString();
          if (sb == null) {
            return simvalue;
          }
          _ = sb.Append(simvalue);
          break;
        case CBORType.FloatingPoint:
          {
            long bits = obj.AsDoubleBits();
            simvalue = bits == DoubleNegInfinity ? "-Infinity" : (
                bits == DoublePosInfinity ? "Infinity" : (
                  CBORUtilities.DoubleBitsNaN(bits) ? "NaN" :
  obj.Untag().ToJSONString()));
            if (sb == null) {
              return simvalue;
            }
            _ = sb.Append(simvalue);
            break;
          }
        case CBORType.ByteString:
          {
            sb = sb ?? new StringBuilder();
            _ = sb.Append("h'");
            byte[] data = obj.GetByteString();
            int length = data.Length;
            for (int i = 0; i < length; ++i) {
              _ = sb.Append(HexAlphabet[(data[i] >> 4) & 15]);
              _ = sb.Append(HexAlphabet[data[i] & 15]);
            }
            _ = sb.Append((char)0x27);
            break;
          }
        case CBORType.TextString:
          {
            sb = sb ?? new StringBuilder();
            _ = sb.Append('\"');
            string ostring = obj.AsString();
            int length = ostring.Length;
            for (int i = 0; i < length; ++i) {
              int cp = DataUtilities.CodePointAt(ostring, i, 0);
              if (cp >= 0x10000) {
                _ = sb.Append("\\U");
                _ = sb.Append(HexAlphabet[(cp >> 20) & 15]);
                _ = sb.Append(HexAlphabet[(cp >> 16) & 15]);
                _ = sb.Append(HexAlphabet[(cp >> 12) & 15]);
                _ = sb.Append(HexAlphabet[(cp >> 8) & 15]);
                _ = sb.Append(HexAlphabet[(cp >> 4) & 15]);
                _ = sb.Append(HexAlphabet[cp & 15]);
                ++i;
              } else if (cp >= 0x7F || cp < 0x20 || cp == '\\' || cp ==
  '\"') {
                _ = sb.Append("\\u");
                _ = sb.Append(HexAlphabet[(cp >> 12) & 15]);
                _ = sb.Append(HexAlphabet[(cp >> 8) & 15]);
                _ = sb.Append(HexAlphabet[(cp >> 4) & 15]);
                _ = sb.Append(HexAlphabet[cp & 15]);
              } else {
                _ = sb.Append((char)cp);
              }
            }
            _ = sb.Append('\"');
            break;
          }
        case CBORType.Array:
          {
            sb = sb ?? new StringBuilder();
            var first = true;
            _ = sb.Append('[');
            if (depth >= 50) {
              _ = sb.Append("...");
            } else {
              for (int i = 0; i < obj.Count; ++i) {
                if (!first) {
                  _ = sb.Append(", ");
                }
                _ = sb.Append(ToStringHelper(obj[i], depth + 1));
                first = false;
              }
            }
            _ = sb.Append(']');
            break;
          }
        case CBORType.Map:
          {
            sb = sb ?? new StringBuilder();
            var first = true;
            _ = sb.Append('{');
            if (depth >= 50) {
              _ = sb.Append("...");
            } else {
              ICollection<KeyValuePair<CBORObject, CBORObject>> entries =
                obj.Entries;
              foreach (KeyValuePair<CBORObject, CBORObject> entry
                in entries) {
                CBORObject key = entry.Key;
                CBORObject value = entry.Value;
                if (!first) {
                  _ = sb.Append(", ");
                }
                _ = sb.Append(ToStringHelper(key, depth + 1));
                _ = sb.Append(": ");
                _ = sb.Append(ToStringHelper(value, depth + 1));
                first = false;
              }
            }
            _ = sb.Append('}');
            break;
          }
        default: {
            sb = sb ?? new StringBuilder();
            _ = sb.Append("???");
            break;
          }
      }
      // Append closing tags if needed
      curobject = obj;
      while (curobject.IsTagged) {
        _ = sb.Append(')');
        curobject = curobject.UntagOne();
      }
      return sb.ToString();
    }

    internal static readonly JSONOptions DefaultOptions =
      new JSONOptions(String.Empty);
    private static readonly JSONOptions PreserveNegZeroNo =
      new JSONOptions("preservenegativezero=0");
    private static readonly JSONOptions PreserveNegZeroYes =
      new JSONOptions("preservenegativezero=1");

    /// <summary>Parses a number whose format follows the JSON
    /// specification. The method uses a JSONOptions with all default
    /// properties.</summary>
    /// <param name='str'>A text string to parse as a JSON number.</param>
    /// <returns>A CBOR object that represents the parsed number. Returns
    /// null if the parsing fails, including if the string is null or
    /// empty.</returns>
    public static CBORObject ParseJSONNumber(string str) {
      return ParseJSONNumber(str, JSONOptions.Default);
    }

    /// <summary>Parses a number whose format follows the JSON
    /// specification (RFC 8259) and converts that number to a CBOR
    /// object.</summary>
    /// <param name='str'>A text string to parse as a JSON number.</param>
    /// <param name='options'>An object containing options to control how
    /// JSON numbers are decoded to CBOR objects. Can be null, in which
    /// case a JSONOptions object with all default properties is used
    /// instead.</param>
    /// <returns>A CBOR object that represents the parsed number. Returns
    /// null if the parsing fails, including if the string is null or
    /// empty.</returns>
    /// <remarks>Roughly speaking, a valid JSON number consists of an
    /// optional minus sign, one or more basic digits (starting with 1 to 9
    /// unless there is only one digit and that digit is 0), an optional
    /// decimal point (".", full stop) with one or more basic digits, and
    /// an optional letter E or e with an optional plus or minus sign and
    /// one or more basic digits (the exponent). A text string representing
    /// a valid JSON number is not allowed to contain white space
    /// characters, including spaces.</remarks>
    public static CBORObject ParseJSONNumber(
      string str,
      JSONOptions options) {
      return String.IsNullOrEmpty(str) ? null :
        ParseJSONNumber(str,
          0,
          str.Length,
          options);
    }

    /// <summary>Parses a number whose format follows the JSON
    /// specification (RFC 8259) from a portion of a text string, and
    /// converts that number to a CBOR object.</summary>
    /// <param name='str'>A text string containing the portion to parse as
    /// a JSON number.</param>
    /// <param name='offset'>An index, starting at 0, showing where the
    /// desired portion of <paramref name='str'/> begins.</param>
    /// <param name='count'>The length, in code units, of the desired
    /// portion of <paramref name='str'/> (but not more than <paramref
    /// name='str'/> 's length).</param>
    /// <returns>A CBOR object that represents the parsed number. Returns
    /// null if the parsing fails, including if the string is null or
    /// empty.</returns>
    /// <exception cref='ArgumentException'>Either <paramref
    /// name='offset'/> or <paramref name='count'/> is less than 0 or
    /// greater than <paramref name='str'/> 's length, or <paramref
    /// name='str'/> 's length minus <paramref name='offset'/> is less than
    /// <paramref name='count'/>.</exception>
    /// <exception cref='ArgumentNullException'>The parameter <paramref
    /// name='str'/> is null.</exception>
    /// <remarks>Roughly speaking, a valid JSON number consists of an
    /// optional minus sign, one or more basic digits (starting with 1 to 9
    /// unless there is only one digit and that digit is 0), an optional
    /// decimal point (".", full stop) with one or more basic digits, and
    /// an optional letter E or e with an optional plus or minus sign and
    /// one or more basic digits (the exponent). A text string representing
    /// a valid JSON number is not allowed to contain white space
    /// characters, including spaces.</remarks>
    public static CBORObject ParseJSONNumber(
      string str,
      int offset,
      int count) {
      return String.IsNullOrEmpty(str) ? null :
        ParseJSONNumber(str,
          offset,
          count,
          JSONOptions.Default);
    }

    internal static CBORObject ParseSmallNumberAsNegative(
      int digit,
      JSONOptions options) {
#if DEBUG
      if (digit <= 0) {
        throw new ArgumentException("digit (" + digit + ") is not greater" +
          "\u0020than 0");
      }
#endif

      if (options != null && options.NumberConversion ==
        JSONOptions.ConversionMode.Double) {
        return CBORObject.FromFloatingPointBits(
           CBORUtilities.IntegerToDoubleBits(-digit),
           8);
      } else if (options != null && options.NumberConversion ==
        JSONOptions.ConversionMode.Decimal128) {
        return CBORObject.FromEDecimal(EDecimal.FromInt32(-digit));
      } else {
        // NOTE: Assumes digit is greater than zero, so PreserveNegativeZeros is
        // irrelevant
        return CBORObject.FromInt32(-digit);
      }
    }

    internal static CBORObject ParseSmallNumber(int digit, JSONOptions
      options) {
#if DEBUG
      if (digit < 0) {
        throw new ArgumentException("digit (" + digit + ") is not greater" +
          "\u0020or equal to 0");
      }
#endif

      if (options != null && options.NumberConversion ==
        JSONOptions.ConversionMode.Double) {
        return CBORObject.FromFloatingPointBits(
           CBORUtilities.IntegerToDoubleBits(digit),
           8);
      } else if (options != null && options.NumberConversion ==
        JSONOptions.ConversionMode.Decimal128) {
        return CBORObject.FromEDecimal(EDecimal.FromInt32(digit));
      } else {
        // NOTE: Assumes digit is nonnegative, so PreserveNegativeZeros is irrelevant
        return CBORObject.FromInt32(digit);
      }
    }

    /// <summary>Parses a number whose format follows the JSON
    /// specification (RFC 8259) and converts that number to a CBOR
    /// object.</summary>
    /// <param name='str'>A text string to parse as a JSON number.</param>
    /// <param name='offset'>An index, starting at 0, showing where the
    /// desired portion of <paramref name='str'/> begins.</param>
    /// <param name='count'>The length, in code units, of the desired
    /// portion of <paramref name='str'/> (but not more than <paramref
    /// name='str'/> 's length).</param>
    /// <param name='options'>An object containing options to control how
    /// JSON numbers are decoded to CBOR objects. Can be null, in which
    /// case a JSONOptions object with all default properties is used
    /// instead.</param>
    /// <returns>A CBOR object that represents the parsed number. Returns
    /// null if the parsing fails, including if the string is null or empty
    /// or <paramref name='count'/> is 0 or less.</returns>
    /// <exception cref='ArgumentNullException'>The parameter <paramref
    /// name='str'/> is null.</exception>
    /// <exception cref='ArgumentException'>Unsupported conversion
    /// kind.</exception>
    /// <remarks>Roughly speaking, a valid JSON number consists of an
    /// optional minus sign, one or more basic digits (starting with 1 to 9
    /// unless there is only one digit and that digit is 0), an optional
    /// decimal point (".", full stop) with one or more basic digits, and
    /// an optional letter E or e with an optional plus or minus sign and
    /// one or more basic digits (the exponent). A text string representing
    /// a valid JSON number is not allowed to contain white space
    /// characters, including spaces.</remarks>
    public static CBORObject ParseJSONNumber(
      string str,
      int offset,
      int count,
      JSONOptions options) {
      return CBORDataUtilitiesTextString.ParseJSONNumber(
        str,
        offset,
        count,
        options,
        null);
    }

    /// <summary>Parses a number from a byte sequence whose format follows
    /// the JSON specification (RFC 8259) and converts that number to a
    /// CBOR object.</summary>
    /// <param name='bytes'>A sequence of bytes to parse as a JSON
    /// number.</param>
    /// <param name='offset'>An index, starting at 0, showing where the
    /// desired portion of <paramref name='bytes'/> begins.</param>
    /// <param name='count'>The length, in code units, of the desired
    /// portion of <paramref name='bytes'/> (but not more than <paramref
    /// name='bytes'/> 's length).</param>
    /// <param name='options'>An object containing options to control how
    /// JSON numbers are decoded to CBOR objects. Can be null, in which
    /// case a JSONOptions object with all default properties is used
    /// instead.</param>
    /// <returns>A CBOR object that represents the parsed number. Returns
    /// null if the parsing fails, including if the byte sequence is null
    /// or empty or <paramref name='count'/> is 0 or less.</returns>
    /// <exception cref='ArgumentNullException'>The parameter <paramref
    /// name='bytes'/> is null.</exception>
    /// <exception cref='ArgumentException'>Unsupported conversion
    /// kind.</exception>
    /// <remarks>Roughly speaking, a valid JSON number consists of an
    /// optional minus sign, one or more basic digits (starting with 1 to 9
    /// unless there is only one digit and that digit is 0), an optional
    /// decimal point (".", full stop) with one or more basic digits, and
    /// an optional letter E or e with an optional plus or minus sign and
    /// one or more basic digits (the exponent). A byte sequence
    /// representing a valid JSON number is not allowed to contain white
    /// space characters, including spaces.</remarks>
    public static CBORObject ParseJSONNumber(
      byte[] bytes,
      int offset,
      int count,
      JSONOptions options) {
      return CBORDataUtilitiesByteArrayString.ParseJSONNumber(
        bytes,
        offset,
        count,
        options,
        null);
    }

    /// <summary>Parses a number from a byte sequence whose format follows
    /// the JSON specification (RFC 8259) and converts that number to a
    /// CBOR object.</summary>
    /// <param name='bytes'>A sequence of bytes to parse as a JSON
    /// number.</param>
    /// <param name='options'>An object containing options to control how
    /// JSON numbers are decoded to CBOR objects. Can be null, in which
    /// case a JSONOptions object with all default properties is used
    /// instead.</param>
    /// <returns>A CBOR object that represents the parsed number. Returns
    /// null if the parsing fails, including if the byte sequence is null
    /// or empty.</returns>
    /// <remarks>Roughly speaking, a valid JSON number consists of an
    /// optional minus sign, one or more basic digits (starting with 1 to 9
    /// unless there is only one digit and that digit is 0), an optional
    /// decimal point (".", full stop) with one or more basic digits, and
    /// an optional letter E or e with an optional plus or minus sign and
    /// one or more basic digits (the exponent). A byte sequence
    /// representing a valid JSON number is not allowed to contain white
    /// space characters, including spaces.</remarks>
    public static CBORObject ParseJSONNumber(
      byte[] bytes,
      JSONOptions options) {
      return (bytes == null || bytes.Length == 0) ? null :
        ParseJSONNumber(bytes,
          0,
          bytes.Length,
          options);
    }

    /// <summary>Parses a number whose format follows the JSON
    /// specification (RFC 8259) from a portion of a byte sequence, and
    /// converts that number to a CBOR object.</summary>
    /// <param name='bytes'>A sequence of bytes to parse as a JSON
    /// number.</param>
    /// <param name='offset'>An index, starting at 0, showing where the
    /// desired portion of <paramref name='bytes'/> begins.</param>
    /// <param name='count'>The length, in code units, of the desired
    /// portion of <paramref name='bytes'/> (but not more than <paramref
    /// name='bytes'/> 's length).</param>
    /// <returns>A CBOR object that represents the parsed number. Returns
    /// null if the parsing fails, including if the byte sequence is null
    /// or empty.</returns>
    /// <exception cref='ArgumentException'>Either <paramref
    /// name='offset'/> or <paramref name='count'/> is less than 0 or
    /// greater than <paramref name='bytes'/> 's length, or <paramref
    /// name='bytes'/> 's length minus <paramref name='offset'/> is less
    /// than <paramref name='count'/>.</exception>
    /// <exception cref='ArgumentNullException'>The parameter <paramref
    /// name='bytes'/> is null.</exception>
    /// <remarks>Roughly speaking, a valid JSON number consists of an
    /// optional minus sign, one or more basic digits (starting with 1 to 9
    /// unless there is only one digit and that digit is 0), an optional
    /// decimal point (".", full stop) with one or more basic digits, and
    /// an optional letter E or e with an optional plus or minus sign and
    /// one or more basic digits (the exponent). A byte sequence
    /// representing a valid JSON number is not allowed to contain white
    /// space characters, including spaces.</remarks>
    public static CBORObject ParseJSONNumber(
      byte[] bytes,
      int offset,
      int count) {
      return (bytes == null || bytes.Length == 0) ? null :
        ParseJSONNumber(bytes,
          offset,
          count,
          JSONOptions.Default);
    }

    /// <summary>Parses a number from a byte sequence whose format follows
    /// the JSON specification. The method uses a JSONOptions with all
    /// default properties.</summary>
    /// <param name='bytes'>A byte sequence to parse as a JSON
    /// number.</param>
    /// <returns>A CBOR object that represents the parsed number. Returns
    /// null if the parsing fails, including if the byte sequence is null
    /// or empty.</returns>
    public static CBORObject ParseJSONNumber(byte[] bytes) {
      return ParseJSONNumber(bytes, JSONOptions.Default);
    }

    /// <summary>Parses a number from a sequence of <c>char</c> s whose
    /// format follows the JSON specification (RFC 8259) and converts that
    /// number to a CBOR object.</summary>
    /// <param name='chars'>A sequence of <c>char</c> s to parse as a JSON
    /// number.</param>
    /// <param name='offset'>An index, starting at 0, showing where the
    /// desired portion of <paramref name='chars'/> begins.</param>
    /// <param name='count'>The length, in code units, of the desired
    /// portion of <paramref name='chars'/> (but not more than <paramref
    /// name='chars'/> 's length).</param>
    /// <param name='options'>An object containing options to control how
    /// JSON numbers are decoded to CBOR objects. Can be null, in which
    /// case a JSONOptions object with all default properties is used
    /// instead.</param>
    /// <returns>A CBOR object that represents the parsed number. Returns
    /// null if the parsing fails, including if the sequence of <c>char</c>
    /// s is null or empty or <paramref name='count'/> is 0 or
    /// less.</returns>
    /// <exception cref='ArgumentNullException'>The parameter <paramref
    /// name='chars'/> is null.</exception>
    /// <exception cref='ArgumentException'>Unsupported conversion
    /// kind.</exception>
    /// <remarks>Roughly speaking, a valid JSON number consists of an
    /// optional minus sign, one or more basic digits (starting with 1 to 9
    /// unless there is only one digit and that digit is 0), an optional
    /// decimal point (".", full stop) with one or more basic digits, and
    /// an optional letter E or e with an optional plus or minus sign and
    /// one or more basic digits (the exponent). A sequence of <c>char</c>
    /// s representing a valid JSON number is not allowed to contain white
    /// space characters, including spaces.</remarks>
    public static CBORObject ParseJSONNumber(
      char[] chars,
      int offset,
      int count,
      JSONOptions options) {
      return CBORDataUtilitiesCharArrayString.ParseJSONNumber(
        chars,
        offset,
        count,
        options,
        null);
    }

    /// <summary>Parses a number from a sequence of <c>char</c> s whose
    /// format follows the JSON specification (RFC 8259) and converts that
    /// number to a CBOR object.</summary>
    /// <param name='chars'>A sequence of <c>char</c> s to parse as a JSON
    /// number.</param>
    /// <param name='options'>An object containing options to control how
    /// JSON numbers are decoded to CBOR objects. Can be null, in which
    /// case a JSONOptions object with all default properties is used
    /// instead.</param>
    /// <returns>A CBOR object that represents the parsed number. Returns
    /// null if the parsing fails, including if the sequence of <c>char</c>
    /// s is null or empty.</returns>
    /// <remarks>Roughly speaking, a valid JSON number consists of an
    /// optional minus sign, one or more basic digits (starting with 1 to 9
    /// unless there is only one digit and that digit is 0), an optional
    /// decimal point (".", full stop) with one or more basic digits, and
    /// an optional letter E or e with an optional plus or minus sign and
    /// one or more basic digits (the exponent). A sequence of <c>char</c>
    /// s representing a valid JSON number is not allowed to contain white
    /// space characters, including spaces.</remarks>
    public static CBORObject ParseJSONNumber(
      char[] chars,
      JSONOptions options) {
      return (chars == null || chars.Length == 0) ? null :
        ParseJSONNumber(chars,
          0,
          chars.Length,
          options);
    }

    /// <summary>Parses a number whose format follows the JSON
    /// specification (RFC 8259) from a portion of a sequence of
    /// <c>char</c> s, and converts that number to a CBOR object.</summary>
    /// <param name='chars'>A sequence of <c>char</c> s to parse as a JSON
    /// number.</param>
    /// <param name='offset'>An index, starting at 0, showing where the
    /// desired portion of <paramref name='chars'/> begins.</param>
    /// <param name='count'>The length, in code units, of the desired
    /// portion of <paramref name='chars'/> (but not more than <paramref
    /// name='chars'/> 's length).</param>
    /// <returns>A CBOR object that represents the parsed number. Returns
    /// null if the parsing fails, including if the sequence of <c>char</c>
    /// s is null or empty.</returns>
    /// <exception cref='ArgumentException'>Either <paramref
    /// name='offset'/> or <paramref name='count'/> is less than 0 or
    /// greater than <paramref name='chars'/> 's length, or <paramref
    /// name='chars'/> 's length minus <paramref name='offset'/> is less
    /// than <paramref name='count'/>.</exception>
    /// <exception cref='ArgumentNullException'>The parameter <paramref
    /// name='chars'/> is null.</exception>
    /// <remarks>Roughly speaking, a valid JSON number consists of an
    /// optional minus sign, one or more basic digits (starting with 1 to 9
    /// unless there is only one digit and that digit is 0), an optional
    /// decimal point (".", full stop) with one or more basic digits, and
    /// an optional letter E or e with an optional plus or minus sign and
    /// one or more basic digits (the exponent). A sequence of <c>char</c>
    /// s representing a valid JSON number is not allowed to contain white
    /// space characters, including spaces.</remarks>
    public static CBORObject ParseJSONNumber(
      char[] chars,
      int offset,
      int count) {
      return (chars == null || chars.Length == 0) ? null :
        ParseJSONNumber(chars,
          offset,
          count,
          JSONOptions.Default);
    }

    /// <summary>Parses a number from a sequence of <c>char</c> s whose
    /// format follows the JSON specification. The method uses a
    /// JSONOptions with all default properties.</summary>
    /// <param name='chars'>A sequence of <c>char</c> s to parse as a JSON
    /// number.</param>
    /// <returns>A CBOR object that represents the parsed number. Returns
    /// null if the parsing fails, including if the sequence of <c>char</c>
    /// s is null or empty.</returns>
    public static CBORObject ParseJSONNumber(char[] chars) {
      return ParseJSONNumber(chars, JSONOptions.Default);
    }
  }
}
